api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+- application state restore logic: on a state restore event, we should be able to figure out which slot contains each spilled register's app value. This is complicated by the fact that registers may be spilled by instrumentation added by multiple phases, and the spill regions may overlap which causes the spilled application value to be moved between spill slots.
+
+We explored the following ideas to avoid spill slot conflicts in drreg:
+
+#### Disjoint slot spaces or arenas
+We can ask drreg to create slot spaces or arenas at init time, which are assigned disjoint spill slots. When reserving a register, the user passes in a “space/arena Id '' to instruct drreg to pick free slots only from that arena. This requires keeping some global drreg state. This also requires the user to guess the best configuration for assigning slots to the arenas, and passing the correct arena Id before each reservation. It may artifically make some spill slots unavailable for use, thereby reducing efficiency.
+
+
+#### Assign phase Id to slots
+Instead of creating slot spaces at init time with a best-guess assignment of slots, we can instead assign a phase Id to slots when they are requested in that phase. We then avoid using slots that are already assigned a phase Id, when we are not in that phase where the slot was used before. This also requires keeping some global drreg state. This does not help in avoiding spill slot conflicts between multiple clients in the same phase.
+
+
+#### Preferred: Scan fragment to determine eligible slots
+When picking a spill slot, we can determine whether using it will cause a slot conflict by scanning for its uses in the current fragment after the current instruction. We pick only that spill slot which does not have any later uses in the current fragment. This does not require any init time guesses or keeping any global drreg state. It does not impose any additional responsibilities on the users, and it also works for multiple clients in the same phase. This was implemented to pick spill slots for GPRs ([commit](https://github.com/DynamoRIO/dynamorio/commit/238bb25de02d4741f52ec368e2118244329d76f0)) and aflags too ([commit](https://github.com/DynamoRIO/dynamorio/commit/a3d0419e4d05113ca8c665f4ef0edb970e3bcf58)).
+
+
+### State Restoration For Drreg
+
+Owner: [Abhinav Sharma](https://github.com/abhinav92003)
+
+Upstream issue: [DynamoRIO/dynamorio#3823](https://github.com/DynamoRIO/dynamorio/issues/3823), [DynamoRIO/dynamorio#3801](https://github.com/DynamoRIO/dynamorio/issues/3801)
+
+On a state restore event, drreg should be able to restore all spilled registers to their application values.
+
+Unfortunately, when a state restore event happens, we only have the encoded fragment, and none of the drreg state, like the reg to spill slot mappings. We need to reconstruct this state based on the faulting pc and the encoded fragment.
```

Comment:
We could store state (have to do this anyway for fragments where recreation is not possible)

Reply:
Probably -- though would we store it for all fragments? That would increase memory usage a lot I think.

Not adding it to this doc though, unless that was what you were suggesting.

Reply:
If I recall correctly, this is not a limitation of the multi-phase drreg design, but only of the implementation. The current restore logic without ilist (`drreg_event_restore_state_without_ilist`) is somewhat stale (doesn't handle multi-phase reservations), but it can be made to be very similar to `drreg_event_restore_state_with_ilist`. The only thing we'll be  missing out on is detection of false spills and restore, which is a general limitation, not related to the scatter/gather sequence.

I have created #5303 for this. Given that it is a bug that we intend to fix (I'll try to find some time soon for this), maybe it's not worth adding it to the design limitation?

Reply:
Agreed, the without_ilist support is fairly good, so typical translation would probably work for drreg after all.  Also xref #4937 for general stored-xl8 support.  OK.